src/libAtomVM/nifs.c

```diff
@@ -4242,6 +4248,36 @@ static term nif_code_load_binary(Context *ctx, int argc, term argv[])
     return result;
 }
 
+static term nif_lists_reverse(Context *ctx, int argc, term argv[])
+{
+    // Compared to erlang version, compute the length of the list and allocate
+    // at once the space for the reverse.
+    int proper;
+    size_t len = term_list_length(argv[0], &proper);
+    if (!proper) {
+        RAISE_ERROR(BADARG_ATOM);
```

Comment:
We should add a very short note here, erlang implementation fails with a function clause error when using a list such as `[1 | 5]` and with bad argument with a list such as `[1, 2, 3 | 5]`.
So I would just mention in a short comment this different behavior.

Reply:
So I did update the documentation for lists:reverse/1.

I'm not sure this detail is important enough in the documentation, but your call. The reason why Erlang/OTP doesn't badarg but goes with function clause for some elements is just because of their optimization to not go into the nif for simple cases.

@fadushin I also reviewed the documentation I added for lists:reverse/2 and especially the discussion on memory impact of tail recursive / non-tail recursive implementation. Does it answer your questions about the few cases that were rewritten? Would you suggest improvements?

https://github.com/atomvm/AtomVM/pull/806/commits/5817b40379d66da0770b97b5d94bba38dafcd4cb#diff-7cbbca808e26538cb5d8a3233b4efdeab0e4dcfeccecebff91cf7824239da4daR149-R163

Reply:
I’m good.  I think I resolved all the open questions I had.